### PR TITLE
Realm js pr workflow

### LINKED ISSUE #3761: Migrate to Github Action
* Validation
  * [x] Linting
  * [x] API doc
* Build steps
  * [x] node/electron (Linux)
  * [x] node/electron (Windows)
  * [x] node/electron (MacOS)
  * [x] react native (iOS)
  * [x] react native (Android)
* Unit testing
  * [x] ~~node (Linux)~~
  * [x] ~~node (Windows)~~
  * [x]: ~~node (MacOS)~~
  * [x] ~~electron (Linux)~~
  * [ ] electron (Windows)
  * [x] ~~electron (MacOS)~~
  * [x] ~~react native (iOS)~~
  * [x] ~~react native (Android)~~
* Integration testing (requires an action to "start sync server")
  * [ ] node (Linux)
  * [ ] node (Windows)
  * [ ] node (MacOS)
  * [ ] electron (Linux)
  * [ ] electron (Windows)
  * [ ] electron (MacOS)
  * [ ] react native (iOS)
  * [ ] react native (Android)
* Example testing
  * [ ] react native (iOS)
  * [ ] react native (Android)

### PR DESCRIPTION
**Pull request by iamer**
## What, How & Why?

This closes #3761

## ☑️ ToDos

* [ ] 📝 Changelog entry
* [ ] 📝 `Compatibility` label is updated or copied from previous entry
* [ ] 🚦 Tests
* [ ] 📱 Check the React Native/other sample apps work if necessary
* [ ] 📝 Public documentation PR created or is not necessary
* [ ] 💥 `Breaking` label has been applied or is not necessary

*If this PR adds or changes public API's:*
* [ ] typescript definitions file is updated
* [ ] jsdoc files updated
* [ ] Chrome debug API is updated if API is available on React Native

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -57,6 +57,11 @@
     "rebuild": "npm run rebuild-changes",
     "build": "npm run build-changes",
     "build:arm": "npm run build-changes:arm",
+    "build:android": "npm ci --ignore-scripts; node scripts/build-android.js",
```
Comment (kraenhansen) [inline package.json:60]: I don't like having `npm ci` (nor `npm install`) in scripts, since this makes the scripts less re-usable. It basically just adds time spent redoing stuff (installing dependencies) if these scripts are called in sequence.

I believe it's better to install dependencies once and explicitly in the workflow.

File: scripts/build-android.js [hunk 2]
```diff
@@ -68,7 +68,7 @@ if (options.clean) {
 }
 
 if (!fs.existsSync(buildPath)) {
-  fs.mkdirSync(buildPath);
+  fs.mkdirSync(buildPath, { recursive: true });
 }
```
Comment (kraenhansen) [inline scripts/build-android.js:72]: Might as well go all the way and make use [`ensureDirSync`](https://github.com/jprichardson/node-fs-extra/blob/HEAD/docs/ensureDir-sync.md):

```
fs.ensureDirSync(buildPath);
```

File: .github/workflows/integration-tests.yml [hunk 3]
```diff
@@ -6,6 +6,7 @@ env:
 
 on:
   pull_request:
+    types: [ labeled ]
```
Comment (kraenhansen) [inline .github/workflows/integration-tests.yml:9]: Why this change?

File: package.json [hunk 4]
```diff
@@ -57,6 +57,11 @@
...
+    "build:ios": "npm ci --ignore-scripts; scripts/build-ios.sh",
+    "build:node-x64": "npm run package -- --arch=x64",
+    "build:node-ia32": "npm run package -- --arch=ia32",
+    "build:node": "npm run package",
```
Comment (kraenhansen) [inline package.json:64]: Why not call `cmake-js` or `prebuilt` directly? It feels that we have multiple script doing almost the same thing.

File: .github/workflows/pr-realm-js.yml [hunk 5]
```diff
@@ -0,0 +1,334 @@
...
+          api-level: 29
+          # ndk: 21.0.6113669
+          # cmake: 3.18.1
+          target: google_apis
+          script: npm run test:android --prefix integration-tests/environments/react-native
+        env:
+          MOCHA_REMOTE_REPORTER: mocha-github-actions-reporter
+          MOCHA_REMOTE_EXIT_ON_ERROR: true
+          SPAWN_LOGCAT: true
+          REALM_BASE_URL:  ${{ secrets.REALM_BASE_URL }}
```
Comment (kraenhansen) [inline .github/workflows/pr-realm-js.yml:302]: It looks [from the logs](https://github.com/realm/realm-js/runs/5003509590?check_suite_focus=true#step:11:22) like this secret isn't set. Ideally this would be an output from the `baas-realm-js` job to ensure we're testing against the same deployment that we created a cluster on.

As a workaround it can be manually set to [the same value that the GitHub Action defaults to](https://github.com/realm/ci-actions/blob/main/mdb-realm/src/helpers.ts#L107):

```suggestion
          REALM_BASE_URL: "https://realm-dev.mongodb.com"
```

File: .github/workflows/pr-realm-js.yml [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+        uses: reactivecircus/android-emulator-runner@v2
+        with:
+          api-level: 29
+          # ndk: 21.0.6113669
+          # cmake: 3.18.1
+          target: google_apis
+          script: npm run test:android --prefix integration-tests/environments/react-native
+        env:
+          MOCHA_REMOTE_REPORTER: mocha-github-actions-reporter
+          MOCHA_REMOTE_EXIT_ON_ERROR: true
```
Comment (kraenhansen) [inline .github/workflows/pr-realm-js.yml:300]:
```suggestion
          MOCHA_REMOTE_EXIT_ON_ERROR: true
          MOCHA_REMOTE_CONTEXT: mongodbRealmBaseUrl=https://realm-dev.mongodb.com
```

File: .github/workflows/integration-tests.yml (hunk 3 above)
Comment (iamer) [reply .github/workflows/integration-tests.yml:9]: this is a temporary change while working on this workflow, to prevent hogging resources.

File: .github/workflows/pr-realm-js.yml [hunk 7]
```diff
@@ -0,0 +1,337 @@
...
+      npm_config_realm_local_prebuilds: ${{github.workspace}}/prebuilds
+      REALM_DISABLE_SYNC_TESTS: true
+      REALM_DISABLE_ANALYTICS: 1
+    strategy:
+      #fail-fast: false
+      matrix:
+        variant:
+          - { os: linux, runner: ubuntu-latest, arch: x64, artifact-path: prebuilds, test-node: true, test-electron: true }
+          - { os: windows, runner: windows-2019, arch: x64, artifact-path: prebuilds, test-node: true, test-electron: true }
+          - { os: windows, runner: windows-2019, arch: ia32, artifact-path: prebuilds }
```
Comment (fronck) [inline .github/workflows/pr-realm-js.yml:18]: Is there any specific reason why we're not using `windows-latest` here?

Comment (iamer) [reply .github/workflows/pr-realm-js.yml:18]: After github's change to use windows 2022 (https://github.com/actions/virtual-environments/issues/4856) for runners, the windows builds were failing.

File: package.json (hunk 4 above)
Comment (iamer) [reply package.json]: Can you please check the latest commits and see if they are clearer now?

File: .github/workflows/pr-realm-js.yml [hunk 8]
```diff
@@ -0,0 +1,397 @@
...
+          privateApiKey: ${{ secrets.ATLAS_PRIVATE_API_KEY }}
+          differentiator: realm js
+
+  integration-tests:
+    name: Integration tests for ${{ matrix.variant.environment }} on ${{ matrix.variant.os }}
+    needs: [build-and-test]
+    #needs: [build-and-test, baas-realm-js]
+    env:
+      REALM_DISABLE_ANALYTICS: 1
+      MOCHA_REMOTE_TIMEOUT: 500000
```
Comment (kraenhansen) [inline .github/workflows/pr-realm-js.yml:249]: Why is this necessary? It seems excessive to me.

File: tests/ReactTestApp/metro.config.js [hunk 9]
```diff
@@ -1,15 +1,162 @@
...
+    blockList: exclusionList(blockList),
+    extraNodeModules,
+  },
   transformer: {
     getTransformOptions: async () => ({
       transform: {
-        experimentalImportSupport: false,
         inlineRequires: true,
       },
     }),
```
Comment (kraenhansen) [inline tests/ReactTestApp/metro.config.js:162]: Could we reuse the [metro-config](https://github.com/realm/realm-js/tree/master/packages/metro-config) package instead of copying it here?

Comment (takameyer) [reply tests/ReactTestApp/metro.config.js:162]: Actually the github action isn't even using the testapp.  I did however make a PR where I attempt to add this:
https://github.com/realm/realm-js/pull/4368

File: Jenkinsfile [hunk 10]
```diff
@@ -1,535 +0,0 @@
-#!groovy
```
Comment (takameyer) [inline Jenkinsfile:1]: We should keep this around for the legacy tests.

**Comment by iamer** [comment]
Already merged through another branch
